tests/test_dirhash.py

```diff
@@ -282,12 +282,45 @@ def test_ignore_hidden_files(self):
         assert filepaths == ['.d2/f1', '.f2', 'd1/.f2', 'd1/f1', 'f1']
 
         # with ignore
-        filepaths = included_paths(
-            self.path_to('root'),
-            match=['*', '!.*']
-        )
+        filepaths = included_paths(self.path_to('root'), match=['*', '!.*'])
+        assert filepaths == ['d1/f1', 'f1']
+
+    def test_ignore_hidden_files_only(self):
+        self.mkdirs('root/d1')
+        self.mkdirs('root/.d2')
+
+        self.mkfile('root/f1')
+        self.mkfile('root/.f2')
+        self.mkfile('root/d1/f1')
+        self.mkfile('root/d1/.f2')
+        self.mkfile('root/.d2/f1')
+
+        # no ignore
+        filepaths = included_paths(self.path_to('root'))
+        assert filepaths == ['.d2/f1', '.f2', 'd1/.f2', 'd1/f1', 'f1']
+
+        # with ignore
+        filepaths = included_paths(self.path_to('root'), match=['**/*', '!**/.*', '**/.*/*', '!**/.*/.*'])
         assert filepaths == ['.d2/f1', 'd1/f1', 'f1']
 
+    def test_ignore_hidden_explicitly_recursive(self):
```

Comment:
3. test - pattern '!**/.*' that leads to the same result as '!.*' (test 1) with the latest pathspec